echopop/biology.py

```diff
@@ -0,0 +1,1212 @@
+    species_id: np.float64
+        Numeric code representing a particular species of interest
+    """
+
+    # If a single dataframe, convert to a list
+    if isinstance(dataframe_list, pd.DataFrame):
+        dataframe_list = [dataframe_list]
+
+    # Filter out the species-of-interest
+    filtered_dataframes = tuple(df[df.species_id == species_id] for df in dataframe_list)
+
+    # Return output
+    return filtered_dataframes
+
+
+def fit_length_weight_relationship(
+    specimen_data: pd.DataFrame, length_distribution: pd.DataFrame
+) -> dict:
+    """
+    Fit a length-weight relationship across discrete bins
+
+    Parameters
+    ----------
+    specimen_data : pd.DataFrame
+        Dataframe containing aged length and weight data.
+    length_distribution: pd.DataFrame
+        Dataframe containing the overall quantized length bins and intervals.
+
+    Notes
+    -----
+    This function first fits a length-weight regression based on measured
+    values and then produces an array of fitted weight values based on
+    binned length values.
+    The length-weight relationship produced here are used for later
+    biomass calculations and apportionment.
+    """
+
+    # Gather specimen measurements to represent 'all' fish
+    specimen_data_all = specimen_data.assign(sex="all")
+
+    # Combine sexed and 'all' specimens
+    # ---- Vertical concatenation
+    specimen_data_all = pd.concat(
+        [specimen_data[specimen_data.group_sex == "sexed"], specimen_data_all]
+    )
+    # ---- Remove bad values
+    specimen_data_all.dropna(subset=["length", "weight"], inplace=True)
+
+    # Fit length-weight linear regression by male, female, and all fish
+    length_weight_regression_df = (
+        specimen_data_all.groupby(["species_id", "sex"])
+        .apply(
+            lambda df: pd.Series(
+                np.polyfit(np.log10(df["length"]), np.log10(df["weight"]), 1),
+                index=["rate", "initial"],
+            ),
+            include_groups=False,
+        )
+        .reset_index()
+    )
+
+    # Predict weights for binned lengths
+    # ---- Initialize dataframe
+    weight_fitted_df = length_distribution.copy()
+    # ---- Expand/merge with length-weight regression coefficients
+    weight_fitted_df = weight_fitted_df.merge(length_weight_regression_df, how="cross")
+    # ---- Predict weight per bin
+    weight_fitted_df["weight_modeled"] = (
+        10.0 ** weight_fitted_df["initial"]
+        * weight_fitted_df["length_bins"] ** weight_fitted_df["rate"]
+    )
+    # ---- Drop unused columns
+    weight_fitted_df = weight_fitted_df.filter(
+        ["length_intervals", "species_id", "sex", "weight_modeled"]
+    )
+
+    # Adjust for cases where there are too few (< 5) specimens within a given length bin
+    # ---- Count number of specimens across length bins
+    weight_fitted_distribution_df = specimen_data_all.count_variable(
```

Comment:
This is not really the right place to mark this, but since `count_variable` is not part of the changed code so I can't comment on it easily.
I wonder if what you are trying to do in `count_variable` is what [pandas groupby value_counts](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.value_counts.html) does